### PR TITLE
REFACTOR-#1934: move MultiIndex checks to backend

### LINKED ISSUE #1934: [REFACTOR] Move MultiIndex checks to backend
Another refactoring to reduce index accesses in API level

### PR DESCRIPTION
**Pull request by ienkovich**
## What do these changes do?
Another refactoring to move index accesses to backend to enable lazy computations

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1934 
- [x] tests passing

### CODE REVIEW

File: modin/backends/base/query_compiler.py [hunk 1]
```diff
@@ -1224,3 +1224,20 @@ def delitem(self, key):
...
+
+    @abc.abstractmethod
+    def has_multiindex(self, axis=0):
+        """
+        Check if specified axis is indexed by MultiIndex.
+
+        Parameters
+        ----------
+        axis : 0 or 1
+            The axis to check.
```
Comment (YarShev) [inline modin/backends/base/query_compiler.py:1236]:
```suggestion
            The axis to check (0 - index, 1 - columns).
```

File: modin/backends/pandas/query_compiler.py [hunk 2]
```diff
@@ -2278,3 +2278,22 @@ def cat_codes(self):
...
 
     # END Cat operations
+
+    def has_multiindex(self, axis=0):
+        """
+        Check if specified axis is indexed by MultiIndex.
+
+        Parameters
+        ----------
+        axis : 0 or 1
```
Comment (YarShev) [inline modin/backends/pandas/query_compiler.py:2288]:
```suggestion
        axis : 0 or 1, default 0
```

File: modin/backends/pandas/query_compiler.py [hunk 3]
```diff
@@ -2278,3 +2278,22 @@ def cat_codes(self):
...
+            The axis to check.
```
Comment (YarShev) [inline modin/backends/pandas/query_compiler.py:2289]:
```suggestion
            The axis to check (0 - index, 1 - columns).
```

**Review by YarShev — CHANGES_REQUESTED** [review]
I saw some more places where we can replace similar checks (both for columns and index) in order to avoid code inconsistencies. Can you please replace the corresponding checks?

**Comment by ienkovich** [comment]
Thanks for your review! I added more checks using the new method and fixed docstrings.

**Comment by YarShev** [comment]
@ienkovich , something went wrong with tests. Try to rebase in order to rerun tests.

File: modin/pandas/base.py [hunk 4]
```diff
@@ -748,7 +748,7 @@ def count(self, axis=0, level=None, numeric_only=False):
             self._validate_dtypes(numeric_only=numeric_only)
 
         if level is not None:
-            if not isinstance(self.axes[axis], pandas.MultiIndex):
+            if not self._query_compiler.has_multiindex(axis):
```
Comment (YarShev) [inline modin/pandas/base.py:751]:
```suggestion
            if not self._query_compiler.has_multiindex(axis=axis):
```

**Review by YarShev — CHANGES_REQUESTED** [review]
@ienkovich , LGTM, but left one more minor comment regarding passing of parameter.

File: modin/backends/pandas/query_compiler.py [hunk 5]
```diff
@@ -1589,10 +1589,7 @@ def sort_index(self, **kwargs):
         sort_remaining = kwargs.pop("sort_remaining", True)
         kwargs["inplace"] = False
 
-        if level is not None or (
-            (axis == 0 and isinstance(self.index, pandas.MultiIndex))
-            or (axis == 1 and isinstance(self.columns, pandas.MultiIndex))
-        ):
+        if level is not None or self.has_multiindex(axis):
```
Comment (YarShev) [inline modin/backends/pandas/query_compiler.py:1592]:
```suggestion
        if level is not None or self.has_multiindex(axis=axis):
```

**Review by YarShev — CHANGES_REQUESTED** [review]

**Comment by YarShev** [comment]
@ienkovich , thanks, LGTM!

**Review by YarShev — APPROVED** [review]

**Pull request merged**